### PR TITLE
Project index export as CSV, XLS

### PR DESCRIPTION
**Pull request by oliverguenther**
- [x] Extract common logic from work package exporters where possible
- Add exporter functionality to Projects
  - [x] Add custom fields handling without columns on projects
  - [x] CSV export
  - [x] XLS export
- [x] Add representation formats to projects API

[OP#37674](https://community.openproject.org/wp/37674)

### CODE REVIEW

File: app/models/exports/formatters/custom_field.rb [hunk 1]
```diff
@@ -0,0 +1,29 @@
+module Exports
+  module Formatters
+    class CustomField < Default
+      ##
+      # Checks if this column is applicable for this column
```
Comment (ulferts) [inline app/models/exports/formatters/custom_field.rb]: With this comment present a couple of times, it should make sense :)

File: lib/api/v3/work_packages/work_package_collection_representer.rb [hunk 2]
```diff
@@ -211,6 +211,9 @@ def representation_formats
             representation_format_pdf_attachments,
             representation_format_pdf_description,
             representation_format_pdf_description_attachments,
+            representation_format_xls,
+            representation_format_xls_descriptions,
+            representation_format_xls_relations,
```
Comment (ulferts) [inline lib/api/v3/work_packages/work_package_collection_representer.rb:216]: 👍 for cleaning that up

**Review by ulferts — CHANGES_REQUESTED** [review]
With the PR having the size it has I focused on testing the functionality which seems to work mostly fine with the following exceptions:

- [x] Sorting of projects
Regardless of the order applied to the project table in the UI, the export will always have a fixed order which is the default order of the table (id + hierarchy):
image

- [ ] The name of the export file is weird containing a random hash (this has been the case before so it can be ignored unless it is a drive by fix)
image

- ~~The export contains archived projects by default. This only changes after the filter has been applied specifically. Because of that, the list deviates from the list in the UI.~~

- [x] The columns "Latest activity at" and "Required disk storage" are always empty

- [x] The column "Public" uses the values true/false while a bool custom fields uses Yes/No. Google tells me that Excel seems to prefer true/false.

- [x] The export seems to suffer from an N+1 query
image

**Comment by oliverguenther** [comment]
>  The export contains archived projects by default. This only changes after the filter has been applied specifically. Because of that, the list deviates from the list in the UI.

I can't reproduce that. If no filters are present in the query, it uses the default set which is active=true

**Comment by ulferts** [comment]
> I can't reproduce that. If no filters are present in the query, it uses the default set which is active=true

You are right. I double checked and it is looking fine. Sorry for the false claim.

**Comment by oliverguenther** [comment]
>  The name of the export file is weird containing a random hash (this has been the case before so it can be ignored unless it is a drive by fix)

That's output from the tempfile that's used to store the export. I added a commit that fixes the filename, but the underlying stored file still has that name and is used by the content-disposition (We have a `filename` column in attachments that's never used apparently). I'd probably want to address this in a subsequent PR and for all attachments if we want to go that way

**Comment by oliverguenther** [comment]
@ulferts I addressed the remarks except for the filename

**Review by ulferts — APPROVED** [review]
👍

**Pull request merged**